File: 2022/java/Day4.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;

class Day4 {
    public static void main(String ... args) throws Exception {
        var input = Files.readString(Path.of("../input/day4.txt"));

        var pairs = input.lines().map(l -> Range.parsePair(l)).toList();

        var sum1 = pairs.stream()
            .filter(pair -> pair[0].fullyContains(pair[1]) || pair[1].fullyContains(pair[0]))
            .count();
        System.out.printf("part 1: %d\n", sum1);

        var sum2 = pairs.stream()
            .filter(pair -> pair[0].overlap(pair[1]) || pair[1].overlap(pair[0]))
            .count();
        System.out.printf("part 2: %d\n", sum2); 
    }
}

record Range (int start, int end) {
    static Range[] parsePair(String line) {
        return Stream.of(line.split(",")).map(s -> parse(s)).toArray(Range[]::new);
    }

    static Range parse(String range) {
        var ends = Stream.of(range.split("-")).mapToInt(s -> Integer.parseInt(s)).toArray();
        return new Range(ends[0], ends[1]);
    }

    boolean fullyContains(Range o) {
        return start <= o.start && end >= o.end;
    }
    
    boolean overlap(Range o) {
        return start <= o.start && end >= o.start;
    }
}

File: 2022/go/day4/day4.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day4.txt")
	lines := strings.Split(string(input), "\n")

	sum1 := 0
	sum2 := 0
	for _, line := range lines {
		sections1, sections2 := parsePair(line)
		if sections1.fullyContains(sections2) || sections2.fullyContains(sections1) {
			sum1 += 1
		}

		if sections1.overlap(sections2) || sections2.overlap(sections1) {
			sum2 += 1
		}
	}
	fmt.Printf("part 1: %v\n", sum1)
	fmt.Printf("part 2: %v\n", sum2)
}

type sections struct {
	start, end int
}

func (s1 sections) fullyContains(s2 sections) bool {
	return s1.start <= s2.start && s1.end >= s2.end
}

func (s1 sections) overlap(s2 sections) bool {
	return s1.start <= s2.start && s1.end >= s2.start
}

func parsePair(line string) (sections, sections) {
	var s1, s2 sections
	fmt.Sscanf(line, "%d-%d,%d-%d", &s1.start, &s1.end, &s2.start, &s2.end)
	return s1, s2
}
